ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1574,6 +1573,52 @@ private StoreFindToken findEntriesFromSegmentStartOffset(Offset initialSegmentSt
+   * value in the index is a ttlUpdate or undelete. In this case, we can make a safe guess that the original blob also
+   * has to be fetched by a replicator anyways and should be counted towards the size limit.
+   * @param entry the {@link JournalEntry} to look up in the index.
+   * @param endOffsetOfSearchRange the end of the range to search in the index.
+   * @param indexSegments the index snapshot to search.
+   * @param messageInfoCache a cache to store previous find results for a key. Queries from an earlier offset will
+   *                         return the same latest value, so there is no need to search the index a second time for
+   *                         keys with multiple journal entries.
+   * @param messageEntries the list of {@link MessageInfo} being constructed. This list will be added to.
+   * @param currentTotalSizeOfEntries a counter for the size in bytes of the entries returned in the query.
+   * @throws StoreException on index search errors.
+   */
+  private void addMessageInfoForJournalEntry(JournalEntry entry, Offset endOffsetOfSearchRange,
+      ConcurrentSkipListMap<Offset, IndexSegment> indexSegments, Map<StoreKey, MessageInfo> messageInfoCache,
+      List<MessageInfo> messageEntries, AtomicLong currentTotalSizeOfEntries) throws StoreException {
+    MessageInfo messageInfo = messageInfoCache.get(entry.getKey());
+    if (messageInfo == null) {
+      // we only need to do an index lookup once per key since the following method will honor any index
+      // values for the key past entry.getOffset()
+      List<IndexValue> valuesInRange =
+          findAllIndexValuesForKey(entry.getKey(), new FileSpan(entry.getOffset(), endOffsetOfSearchRange),
+              EnumSet.allOf(IndexEntryType.class), indexSegments);
+
+      IndexValue latestValue = valuesInRange.get(0);
+      if (valuesInRange.size() > 1 && (latestValue.isTtlUpdate() || latestValue.isUndelete())) {
+        IndexValue earliestValue = valuesInRange.get(valuesInRange.size() - 1);
+        if (earliestValue.isPut() && earliestValue.getOffset().compareTo(entry.getOffset()) >= 0) {
+          currentTotalSizeOfEntries.addAndGet(earliestValue.getSize());
+        }
+      }
+      messageInfo =
+          new MessageInfo(entry.getKey(), latestValue.getSize(), latestValue.isDelete(), latestValue.isTtlUpdate(),
+              latestValue.isUndelete(), latestValue.getExpiresAtMs(), null, latestValue.getAccountId(),
+              latestValue.getContainerId(), latestValue.getOperationTimeInMs(), latestValue.getLifeVersion());
+      messageInfoCache.put(entry.getKey(), messageInfo);
+    }
+    // We may add duplicate MessageInfos to the list here since the ordering of the list is used to calculate
+    // bytes read for the token. The duplicates will be cleaned up by eliminateDuplicates.
+    currentTotalSizeOfEntries.addAndGet(messageInfo.getSize());
```

Comment:
Logic looks good. I just need some clarification on PUT. If the entry is a PUT, how can we avoid its size being added `currentTotalSizeOfEntries` twice ?

Reply:
Sorry, I missed this question this morning. If the most recent entry is a put, the check on line 1604 will be false, so the size will only be accounted for once. The only case where extra size will be added if there are more than one entry found past `entry.getOffset()` and the latest entry is a TTL update or undelete and the earliest value at or past the journal entry is a put (e.g. not something else that doesn't indicate a need for blob fetching like a delete).

Reply:
got it